### PR TITLE
Rework Tabs to follow our latest conventions, fill missing gaps, etc.,

### PR DESCRIPTION
**Pull request by ynotdraw**
## 🚀 Description

Tabs was reworked in many ways to fall in line with other existing components.

- Tab Group is now `tab-group` instead of `tab.group` to follow our naming conventions
- Renamed Tab and Panel to follow our naming conventions - they're now proper children of `tab-group`
- Tab no longer has a default slot - a `label` is exposed instead.
- All internal state has been marked as `private`.
- `name` defaults to `undefined` instead of any empty string.
- Panel's `privateIsSelected` is `privateSelected`.
- Added a "With Icon" story.
- Tab Group now throws if a Tab is added without a corresponding Tab Panel and vice versa.
- Disabled overflow buttons now have a `not-allowed` cursor.
- Design requested a few minor tweaks - the selected indicator is rounded on the top-left and top-right. The transitions use our animation variables.

## 📋 Checklist

- I have followed the [Contributing Guidelines](https://github.com/crowdstrike/glide-core/blob/main/CONTRIBUTING.md).
- I have added tests to cover new or updated functionality.
- I have added or updated Storybook stories.
- I have [localized](https://github.com/CrowdStrike/glide-core/blob/main/CONTRIBUTING.md#translations-and-static-strings) new strings.
- I have followed the [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/) or met with the Accessibility Team.
- I have included a [changeset](https://github.com/CrowdStrike/glide-core/blob/main/CONTRIBUTING.md#versioning-a-package).
- I have scheduled a design review.
- I have reviewed the Storybook and Visual Test Report links below.

## 🔬 Manual Testing

Tests should have us covered for the most part, but it may be wise to pull down this branch and poke around, especially since we don't have GitHub Action access anymore.

Here are some edge cases I made sure we cover, but you can help verify too:

**Tab renames**
- Open Tab Group's story
- Select Tab "Two"
- Update Tab "One"'s label to something longer
- Now delete characters from Tab One's label
- Ensure the overflow buttons don't appear

**Tab icon slot deletions**
- Open Tab Group's story
- Delete Tab "One"'s icon
- Verify the selected tab indicator is the correct size

**Long tab names**
This came from https://github.com/CrowdStrike/glide-core/pull/1141

- Update the existing story to have two Tabs
  - Give the first a label like "Det"
  - Give the second a label like "Minimum Remediations"
- Select the second tab
- Now select the first tab
- Verify the overflow buttons are never shown at any point

### CODE REVIEW

File: src/tab-group.ts [hunk 1]
```diff
@@ -407,50 +442,70 @@ export default class TabGroup extends LitElement {
...
+    }
 
-      this.#selectedTabIndicatorElementRef.value.style.setProperty(
-        '--private-selected-tab-indicator-width',
-        `${selectedTabWidth - selectedTabInlinePadding}px`,
-      );
+    for (const panel of this.#panelElements) {
+      if (!this.#tabElements.some((tab) => tab.panel === panel.name)) {
+        throw new Error(`Panel with name="${panel.name}" has no matching Tab.`);
+      }
```
Comment (ynotdraw) [inline src/tab-group.ts:508]: > Tab Group now throws if a Tab is added without a corresponding Tab Panel and vice versa.

File: src/tab-group.ts [hunk 2]
```diff
@@ -407,50 +442,70 @@ export default class TabGroup extends LitElement {
...
-            window
-              .getComputedStyle(this.#lastSelectedTab)
-              .getPropertyValue('padding-inline-start'),
-          )
-        : isLastTab
+    // We need our measurements and positioning to happen before painting, but after
+    // transitioning the font-weight of the selected tab. Requesting an animation frame
+    // helps with the timing. Otherwise the selected tab indicator won't have the
+    // correct width or positioning.
+    requestAnimationFrame(() => {
```
Comment (ynotdraw) [inline src/tab-group.ts:449]: No changes here other than wrapping it in a `requestAnimationFrame()`

File: src/tab-group.ts [hunk 3]
```diff
@@ -311,49 +328,70 @@ export default class TabGroup extends LitElement {
...
+    }
+  }
+
+  #onTabListFocusOut() {
+    // Set the last selected tab as tabbable so that when pressing Shift + Tab on the
+    // Tab Panel focus goes back to the last selected tab.
+    for (const tabElement of this.#tabElements.values()) {
+      tabElement.tabIndex = tabElement === this.#lastSelectedTab ? 0 : -1;
+    }
+  }
```
Comment (ynotdraw) [inline src/tab-group.ts:390]: Renamed from `#onTabListFocusout()`.

The for loop used to read `for (const [, tabElement] of this.#tabElements.entries()) {`, but `entries()` isn't needed because we only wanted the tabElement. Clearer to use `.values()`.

File: src/tab-group.test.basics.ts [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        ></glide-core-tab-group-tab>
+        <glide-core-tab-group-panel name="1">One</glide-core-tab-group-panel>
+
+        <div slot="nav">Tab 1</div>
+      </glide-core-tab-group>
+    `);
+  });
+});
+
+it('throws when a Tab does not have a corresponding Panel', async () => {
```
Comment (ynotdraw) [inline src/tab-group.test.basics.ts:310]: Here and below are new.

File: src/tab-group.test.basics.ts
Comment (ynotdraw) [inline src/tab-group.test.basics.ts:1]: Quite the bummer. Really only renamed this file but git botched it. I commented below where new tests were added, but everything else was existing code.

File: src/tab-group.tab.ts [hunk 5]
```diff
@@ -56,16 +76,24 @@ export default class Tab extends LitElement {
...
+        new Event('selected', {
+          bubbles: true,
+          composed: true,
+        }),
+      );
+    } else if (hasChanged) {
       this.dispatchEvent(
-        new Event('private-selected', {
+        new Event('deselected', {
+          // No need to be composed, as consumers won't use this event directly.
```
Comment (ynotdraw) [inline src/tab-group.tab.ts:96]: Matches [Option](https://github.com/CrowdStrike/glide-core/blob/main/src/option.ts#L175-L194).

The `selected` event here is composed, as consumers want to know when a tab is selected. Rather than commenting there, I decided to comment on why `deselected` is **not** composed. Let me know if that's odd.

File: src/tab.group.test.events.ts [hunk 6]
```diff
@@ -1,24 +1,30 @@
...
+        slot="nav"
+      ></glide-core-tab-group-tab>
+      <glide-core-tab-group-tab
+        label="Two"
+        panel="2"
+        slot="nav"
+      ></glide-core-tab-group-tab>
 
-      <glide-core-tab-panel name="1">One</glide-core-tab-panel>
-      <glide-core-tab-panel name="1">Two</glide-core-tab-panel>
```
Comment (ynotdraw) [inline src/tab.group.test.events.ts:15]: Actually caught some validation issues of a matching tab + panel here and in a few other tests🎉

File: src/tab-group.test.interactions.ts
Comment (ynotdraw) [inline src/tab-group.test.interactions.ts:1]: Similar story as above with tests, but mostly just renames

File: src/tab-group.ts [hunk 7]
```diff
@@ -54,7 +54,11 @@ export default class TabGroup extends LitElement {
       @keydown=${this.#onComponentKeydown}
       ${ref(this.#componentElementRef)}
     >
-      <div class="tab-container" data-test="tab-container">
+      <div
+        class="tab-container"
+        data-test="tab-container"
+        ${onResize(this.#onTabContainerResize.bind(this))}
```
Comment (ynotdraw) [inline src/tab-group.ts:60]: Moved the resize handling to the container instead. Doing so ensures that `#onTabContainerResize()` only gets called on an actual resize of the element. Either by dragging your window, or constraining the width of the TabGroup.

Before this would trigger in many other cases, such as updating the Tab label, as that's _technically_ a resize. But it's pretty confusing that happens unless you know about it. I think this is more explicit and clear, especially with the other changes.

File: src/tab-group.ts [hunk 8]
```diff
@@ -80,16 +84,18 @@ export default class TabGroup extends LitElement {
           data-test="tablist"
           role="tablist"
           tabindex="-1"
-          @focusout=${this.#onTabListFocusout}
+          @focusout=${this.#onTabListFocusOut}
```
Comment (ynotdraw) [inline src/tab-group.ts:87]:
```diff
- onTabListFocusout
+ onTabListFocusOut
```

File: src/tab-group.ts [hunk 9]
```diff
@@ -311,49 +328,70 @@ export default class TabGroup extends LitElement {
...
+      this.#updateSelectedTabIndicator();
+    }
+
+    for (const panel of this.#panelElements) {
+      panel.privateSelected = panel.name === this.#lastSelectedTab?.panel;
+      panel.tabIndex = panel.name === this.#lastSelectedTab?.panel ? 0 : -1;
+    }
+  }
+
+  #onTabIconSlotOrLabelChange() {
```
Comment (ynotdraw) [inline src/tab-group.ts:357]: This and the changes to `#updateSelectedTabIndicator()` took up most of the time for this PR to get right 😓. Think it's in a good spot now. Let me know if the comments can be improved.

File: src/tab-group.ts [hunk 10]
```diff
@@ -311,49 +331,70 @@ export default class TabGroup extends LitElement {
...
-  #onTabListFocusout() {
-    // Set the last selected tab as tabbable so that when pressing Shift + Tab on the
-    // Tab Panel focus goes back to the last selected tab.
-    for (const [, tabElement] of this.#tabElements.entries()) {
-      tabElement.tabIndex = tabElement === this.#lastSelectedTab ? 0 : -1;
-    }
-  }
-
-  #onTabListResize() {
+  #onTabContainerResize() {
```
Comment (ynotdraw) [inline src/tab-group.ts:334]: This is simplified now - we only need to concern ourselves with the overflow buttons on resize.  Not updating the selected tab indicator.

File: src/tab.ts [hunk 11]
```diff
@@ -81,14 +109,6 @@ export default class Tab extends LitElement {
   @property({ reflect: true })
   override readonly role = 'tab';
 
-  privateSelect() {
```
Comment (ynotdraw) [inline src/tab.ts:84]: Removed `privateSelect()` in favor of Tab Group setting `selected` directly instead.

File: src/tab-group.tab.ts [hunk 12]
```diff
@@ -28,23 +29,42 @@ declare global {
...
+
+  set label(label) {
+    this.#label = label;
+
+    this.dispatchEvent(
+      new Event('private-label-change', {
+        bubbles: true,
+      }),
+    );
+  }
```
Comment (ynotdraw) [inline src/tab-group.tab.ts:66]: > The default slot of Tab has been removed in favor of a `label` attribute.

File: src/tab-group.ts [hunk 13]
```diff
@@ -270,6 +288,7 @@ export default class TabGroup extends LitElement {
 
   #onDefaultSlotChange() {
     this.#setAriaAttributes();
+    this.#validateTabPanelPairs();
```
Comment (ynotdraw) [inline src/tab-group.ts:291]: Panels are added to the default slot, so we need to validate them when it changes.

File: .changeset/stupid-items-greet.md [hunk 14]
```diff
@@ -0,0 +1,10 @@
+---
+'@crowdstrike/glide-core': minor
+---
+
+- The default slot of Tab has been removed in favor of a `label` attribute.
+- Tab Group now throws if a Tab is added without a corresponding Tab Panel and vice versa.
+- `glide-core-tab-panel` has been renamed to `glide-core-tab-group-panel`.
+- The import of `@crowdstrike/glide-core/tab.panel.js` has been updated to `@crowdstrike/glide-core/tab-group.panel.js`.
+- `glide-core-tab` has been renamed to `glide-core-tab-group-tab`.
+- The import of `@crowdstrike/glide-core/tab.js` has been updated to `@crowdstrike/glide-core/tab-group.tab.js`.
```
Comment (clintcs) [inline .changeset/stupid-items-greet.md:10]: Which import are you referring to? I don't think people will know what this means. Tab's _import_ hasn't changed. Instead, Tab has been renamed and imports of it will have to be updated, no?

**Review by danwenzel — APPROVED** [review]

File: src/tab-group.tab.ts [hunk 15]
```diff
@@ -56,16 +76,24 @@ export default class Tab extends LitElement {
    */
   @property({ type: Boolean, reflect: true })
   get selected(): boolean {
-    return this.#isSelected;
+    return this.#selected;
```
[the remark this answers is not in the document]
Comment (ynotdraw) [reply src/tab-group.tab.ts]: https://github.com/CrowdStrike/glide-core/pull/1081/commits/bed7546e68dc0e4e9ecba64fa541f39e76b1fb6e

File: .changeset/stupid-items-greet.md (hunk 14 above)
Comment (ynotdraw) [reply .changeset/stupid-items-greet.md]: https://github.com/CrowdStrike/glide-core/pull/1081/commits/9399f800a3f771881a6ed3c5b657068407840510

File: src/tab-group.ts [hunk 16]
```diff
@@ -408,7 +409,6 @@ export default class TabGroup extends LitElement {
 
   #onTabSelected(event: Event) {
     if (event.target instanceof TabGroupTab) {
-      event.target.selected = true;
```
Comment (ynotdraw) [inline src/tab-group.ts]: The tab already sets selected when clicked or on key down below, so no need to set it here yet again.
